File: tic_tac_zig/game.zig
const std = @import("std");
const Utils = @import("utils.zig");
const Player = @import("player.zig").Player;
const Board = @import("board.zig").Board;

pub const Game = struct {
    player1: Player,
    player2: Player,
    currentPlayer: ?Player,
    winner: ?Player,
    board: Board,

    pub fn init(p1: Player, p2: Player) Game {
        return Game {
            .player1 = p1,
            .player2 = p2,
            .currentPlayer = p1,
            .winner = null,
            .board = Board.init(),
        };
    }

    pub fn switchCurrentPlayer(self: Game) !void {
        if (self.currentPlayer == null) {
            @panic("Cannot find current player");
        }

        if (self.currentPlayer.? == self.player1) {
            self.currentPlayer = self.player2;
        } else if (self.currentPlayer.? == self.player2) {
            self.currentPlayer = self.player1;
        }
    }

    pub fn display(self: Game) void {
        self.board.display();
    }

    pub fn isWinner(self: Game) bool {
        return self.board.checkWinner();
    }

    pub fn isDraw(self: Game) bool {
        return self.board.isFull();
    }

    pub fn isOver(self: Game) bool {
        const out = std.io.getStdOut().writer();
        if (self.isWinner()) {
            self.winner = self.currentPlayer.?;
            self.board.setRemainingCells();
            Utils.clearTerminal();
            self.display();
            try out.print("Game over! ${s} is the winner!\n", .{self.currentPlayer.?.name});
            return true;
        }

        if (self.isDraw()) {
            self.winner = null;
            self.currentPlayer = null;
            try out.print("Game over! It is a draw!\n", .{});
            return true;
        }

        return false;
    }

    pub fn makeMove(self: Game) void {
        const out = std.io.getStdOut().writer();

        if (self.currentPlayer == null) {
            @panic("Cannot find current player");
        }

        while (true) {
            const coords: Utils.Coords = try Utils.getUserCoords(self.currentPlayer.?.name);
            
            if (try self.board.setCell(coords.row, coords.col, self.currentPlayer.?.mark)) |res| {
                if (res == true) {
                    break;
                }
            } else {
                try out.print("Error while making move.", .{});
            }
        }
    }
};

File: tic_tac_deno/game.ts
import Board from "./board.ts";
import Player from "./player.ts";
import { Errors, clearTerminal, getUserCoords } from "./utils.ts";

export default class Game {
    player1: Player
    player2: Player
    currentPlayer: Player | null
    winner: Player | null
    board: Board

    constructor(player1: Player, player2: Player) {
        this.player1 = player1
        this.player2 = player2
        this.currentPlayer = player1
        this.winner = null
        this.board = new Board()
    }

    switchCurrentPlayer(): void {
        if (!this.currentPlayer) {
            throw new Error(Errors.ERR_NO_CURR_PLAYER)
        }

        if (this.currentPlayer == this.player1) {
            this.currentPlayer = this.player2
        } else if (this.currentPlayer == this.player2) {
            this.currentPlayer = this.player1
        }
    }

    display(): void {
        this.board.display()
    }

    isWinner(): boolean {
        return this.board.checkWinner()
    }

    isDraw(): boolean {
        return this.board.isFull()
    }

    isOver(): boolean {
        if (this.isWinner()) {
            this.winner = this.currentPlayer
            this.board.setRemainingCells()
            clearTerminal()
            this.display()
            console.log(`Game over! ${this.currentPlayer?.name} is the winner!`)
            return true
        }
        if (this.isDraw()) {
            this.winner = null
            this.currentPlayer = null
            console.log("Game over! It is a draw!")
            return true
        }
        return false
    }

    makeMove(): void {
        if (!this.currentPlayer) {
            throw new Error(Errors.ERR_NO_CURR_PLAYER)
        }

        while (true) {
            try {
                const [row, col] = getUserCoords(this.currentPlayer.name)
                this.board.setCell(row, col, this.currentPlayer.mark)
                break
            } catch (e: unknown) {
                console.log((e as Error).message)
            }
        }
    }
}